File: skills/skill-reviewer/scripts/skill_eval_measurement.py
#!/usr/bin/env python3
"""Pure measurement-validity rules for executable Skill evals.

This module intentionally knows nothing about workspaces, agents, or evolution
state.  It answers the smaller question that must be settled first: does the
declared deterministic oracle distinguish known-good output from known-bad
output, and is the selected sampling policy internally coherent?
"""

from __future__ import annotations

import re
from typing import Any


TEXT_ASSERTION_TYPES = {
    "text_contains",
    "text_not_contains",
    "text_matches",
    "text_not_matches",
}
CALIBRATION_FIELDS = {"pass_examples", "fail_examples"}


def evaluate_text_assertion(assertion: dict[str, Any], content: str) -> bool:
    """Apply the exact runtime text predicate to one calibration example."""

    assertion_type = assertion.get("type")
    if assertion_type in {"text_contains", "text_not_contains"}:
        raw_expected = assertion.get("expected")
        expected = [raw_expected] if isinstance(raw_expected, str) else raw_expected
        if not isinstance(expected, list) or not all(
            isinstance(value, str) for value in expected
        ):
            raise ValueError("expected must be a string or string array")
        contains_all = all(value in content for value in expected)
        return (
            contains_all
            if assertion_type == "text_contains"
            else not any(value in content for value in expected)
        )
    if assertion_type in {"text_matches", "text_not_matches"}:
        pattern = assertion.get("pattern")
        if not isinstance(pattern, str) or not pattern:
            raise ValueError("pattern must be a non-empty string")
        matched = re.search(pattern, content, flags=re.MULTILINE) is not None
        return matched if assertion_type == "text_matches" else not matched
    raise ValueError(f"unsupported text assertion type: {assertion_type}")


def calibrate_assertion(assertion: dict[str, Any]) -> dict[str, Any]:
    """Return a content-free calibration result for one text assertion."""

    assertion_id = str(assertion.get("id", ""))
    if assertion.get("type") not in TEXT_ASSERTION_TYPES:
        return {
            "assertion_id": assertion_id,
            "status": "not_applicable",
            "pass_example_count": 0,
            "fail_example_count": 0,
            "failed_pass_examples": [],
            "failed_fail_examples": [],
        }
    calibration = assertion.get("calibration")
    if not isinstance(calibration, dict):
        return {
            "assertion_id": assertion_id,
            "status": "unverified",
            "pass_example_count": 0,
            "fail_example_count": 0,
            "failed_pass_examples": [],
            "failed_fail_examples": [],
        }
    pass_examples = calibration.get("pass_examples", [])
    fail_examples = calibration.get("fail_examples", [])
    failed_pass = [
        index
        for index, example in enumerate(pass_examples)
        if not evaluate_text_assertion(assertion, example)
    ]
    failed_fail = [
        index
        for index, example in enumerate(fail_examples)
        if evaluate_text_assertion(assertion, example)
    ]
    return {
        "assertion_id": assertion_id,
        "status": "invalid" if failed_pass or failed_fail else "valid",
        "pass_example_count": len(pass_examples),
        "fail_example_count": len(fail_examples),
        "failed_pass_examples": failed_pass,
        "failed_fail_examples": failed_fail,
    }

def assess_oracle(assertions: list[dict[str, Any]]) -> dict[str, Any]:
    """Aggregate must-pass text calibration without exposing example text."""

    checks = [
        calibrate_assertion(assertion)
        for assertion in assertions
        if assertion.get("severity", "must_pass") == "must_pass"
        and assertion.get("type") in TEXT_ASSERTION_TYPES
    ]
    if any(check["status"] == "invalid" for check in checks):
        status = "invalid"
    elif any(check["status"] == "unverified" for check in checks):
        status = "unverified"
    else:
        status = "valid"
    reasons: list[str] = []
    for check in checks:
        if check["status"] == "invalid":
            reasons.append(f"assertion_calibration_failed:{check['assertion_id']}")
        elif check["status"] == "unverified":
            reasons.append(f"assertion_calibration_missing:{check['assertion_id']}")
    return {
        "status": status,
        "required_text_assertions": len(checks),
        "calibrated_text_assertions": sum(
            check["status"] == "valid" for check in checks
        ),
        "checks": checks,
        "reasons": reasons,
    }


def normalize_sampling(
    raw: object,
    *,
    legacy_repeats: int,
    determinism: str,
) -> dict[str, Any]:
    """Separate sampling from the legacy output-variability classification."""

    if raw is None:
        return {
            "repeats": legacy_repeats,
            "pairing": "paired",
            "source": "legacy-determinism",
        }
    if not isinstance(raw, dict):
        raise ValueError("sampling must be an object")
    unknown = sorted(set(raw) - {"repeats", "pairing"})
    if unknown:
        raise ValueError("sampling contains unsupported fields: " + ", ".join(unknown))
    repeats = raw.get("repeats")
    if (
        not isinstance(repeats, int)
        or isinstance(repeats, bool)
        or repeats < 1
        or repeats > 10
    ):
        raise ValueError("sampling.repeats must be an integer from 1 to 10")
    pairing = raw.get("pairing", "paired")
    if pairing != "paired":
        raise ValueError("sampling.pairing must be paired")
    if determinism == "stochastic" and repeats < 3:
        raise ValueError("stochastic evals require at least three sampling repeats")
    return {"repeats": repeats, "pairing": pairing, "source": "explicit"}


def assess_runtime_measurement(
    *,
    oracle: dict[str, Any],
    sampling: dict[str, Any],
    direction_disagreement: bool,
) -> dict[str, Any]:
    """Combine frozen-oracle and observed paired-sampling validity."""

    oracle_status = str(oracle.get("status", "unverified"))
    sampling_status = "invalid" if direction_disagreement else "valid"
    reasons = list(oracle.get("reasons", []))
    if direction_disagreement:
        reasons.append("paired_sampling_direction_disagreement")
    if "invalid" in {oracle_status, sampling_status}:
        status = "invalid"
    elif oracle_status != "valid":
        status = "unverified"
    else:
        status = "valid"
    return {
        "status": status,
        "oracle": oracle,
        "sampling": {
            "status": sampling_status,
            "repeats": sampling.get("repeats"),
            "pairing": sampling.get("pairing"),
            "source": sampling.get("source"),
            "direction_disagreement": direction_disagreement,
        },
        "reasons": sorted(set(str(reason) for reason in reasons)),
    }

File: skills/skill-reviewer/scripts/lib/skill-eval-measurement.mjs
export const TEXT_ASSERTION_TYPES = new Set([
  "text_contains",
  "text_not_contains",
  "text_matches",
  "text_not_matches",
]);
export const CALIBRATION_FIELDS = new Set(["pass_examples", "fail_examples"]);
export const PORTABLE_REGEX_CONTRACT = "skill-reviewer.ecmascript-regexp-subset.v1";

export function repeatMetricValue(record, metric) {
  if (record === null || typeof record !== "object" || Array.isArray(record)) {
    return null;
  }
  const metrics = record.metrics;
  const value = metric === "required_pass_rate"
    ? record.required_pass_rate
    : metrics !== null && typeof metrics === "object" && !Array.isArray(metrics)
      ? metrics[metric]
      : null;
  return typeof value === "number" && Number.isFinite(value) ? value : null;
}

export function pairedRepeatMetrics({ candidate, baseline, metric }) {
  const candidateRepeats = Array.isArray(candidate?.repeats) ? candidate.repeats : [];
  const baselineRepeats = Array.isArray(baseline?.repeats) ? baseline.repeats : [];
  if (candidateRepeats.length === 0 || candidateRepeats.length !== baselineRepeats.length) {
    return null;
  }
  const baselineByRepeat = new Map();
  for (const record of baselineRepeats) {
    if (
      record === null || typeof record !== "object" || Array.isArray(record)
      || !Number.isInteger(record.repeat) || baselineByRepeat.has(record.repeat)
    ) return null;
    baselineByRepeat.set(record.repeat, record);
  }
  const seenCandidateRepeats = new Set();
  const pairs = [];
  for (const record of candidateRepeats) {
    if (
      record === null || typeof record !== "object" || Array.isArray(record)
      || !Number.isInteger(record.repeat) || seenCandidateRepeats.has(record.repeat)
    ) return null;
    seenCandidateRepeats.add(record.repeat);
    const baselineRecord = baselineByRepeat.get(record.repeat);
    const candidateValue = repeatMetricValue(record, metric);
    const baselineValue = repeatMetricValue(baselineRecord, metric);
    if (candidateValue === null || baselineValue === null) return null;
    pairs.push({ repeat: record.repeat, candidate: candidateValue, baseline: baselineValue });
  }
  if (seenCandidateRepeats.size !== baselineByRepeat.size) return null;
  return pairs.sort((left, right) => left.repeat - right.repeat);
}

function rejectNonPortableEscapes(source) {
  const unsupported = new Set(["A", "Z", "w", "W", "d", "D", "b", "B", "U", "N", "p", "P", "c", "k"]);
  for (let index = 0; index < source.length; index += 1) {
    if (source[index] !== "\\") continue;
    let slashCount = 1;
    while (source[index + slashCount] === "\\") slashCount += 1;
    if (slashCount % 2 === 1) {
      const escaped = source[index + slashCount];
      if (unsupported.has(escaped) || /[1-9]/.test(escaped ?? "")) {
        throw new Error(`non-portable escape \\${escaped} is not supported`);
      }
    }
    index += slashCount - 1;
  }
}

export function compilePortableRegex(pattern, baseFlags = "") {
  const inline = /^\(\?([ims]+)\)/.exec(pattern);
  const source = inline ? pattern.slice(inline[0].length) : pattern;
  if (/\(\?(?:P[<=]|<|>|\(|#|[ims]+[:)])/.test(source)) {
    throw new Error("non-portable group or inline-flag construct is not supported");
  }
  rejectNonPortableEscapes(source);
  const flags = new Set(baseFlags);
  for (const flag of inline?.[1] ?? "") flags.add(flag);
  return new RegExp(source, [...flags].join(""));
}

export function evaluateTextAssertion(assertion, content) {
  if (["text_contains", "text_not_contains"].includes(assertion.type)) {
    const expected = typeof assertion.expected === "string"
      ? [assertion.expected]
      : assertion.expected;
    if (!Array.isArray(expected) || expected.some((value) => typeof value !== "string")) {
      throw new Error("expected must be a string or string array");
    }
    if (assertion.type === "text_contains") {
      return expected.every((value) => content.includes(value));
    }
    return !expected.some((value) => content.includes(value));
  }
  if (["text_matches", "text_not_matches"].includes(assertion.type)) {
    if (typeof assertion.pattern !== "string" || assertion.pattern === "") {
      throw new Error("pattern must be a non-empty string");
    }
    const matched = compilePortableRegex(assertion.pattern, "m").test(content);
    return assertion.type === "text_matches" ? matched : !matched;
  }
  throw new Error(`unsupported text assertion type: ${assertion.type}`);
}

export function calibrateAssertion(assertion) {
  const assertionId = String(assertion.id ?? "");
  const empty = {
    assertion_id: assertionId,
    pass_example_count: 0,
    fail_example_count: 0,
    failed_pass_examples: [],
    failed_fail_examples: [],
  };
  if (!TEXT_ASSERTION_TYPES.has(assertion.type)) {
    return { ...empty, status: "not_applicable" };
  }
  if (!assertion.calibration || typeof assertion.calibration !== "object" || Array.isArray(assertion.calibration)) {
    return { ...empty, status: "unverified" };
  }
  const passExamples = assertion.calibration.pass_examples ?? [];
  const failExamples = assertion.calibration.fail_examples ?? [];
  const failedPass = passExamples
    .map((example, index) => evaluateTextAssertion(assertion, example) ? null : index)
    .filter((index) => index !== null);
  const failedFail = failExamples
    .map((example, index) => evaluateTextAssertion(assertion, example) ? index : null)
    .filter((index) => index !== null);
  return {
    assertion_id: assertionId,
    status: failedPass.length > 0 || failedFail.length > 0 ? "invalid" : "valid",
    pass_example_count: passExamples.length,
    fail_example_count: failExamples.length,
    failed_pass_examples: failedPass,
    failed_fail_examples: failedFail,
  };
}

export function assessOracle(assertions) {
  const checks = assertions
    .filter((assertion) => (assertion.severity ?? "must_pass") === "must_pass")
    .filter((assertion) => TEXT_ASSERTION_TYPES.has(assertion.type))
    .map(calibrateAssertion);
  const status = checks.some((check) => check.status === "invalid")
    ? "invalid"
    : checks.some((check) => check.status === "unverified")
      ? "unverified"
      : "valid";
  const reasons = checks.flatMap((check) => {
    if (check.status === "invalid") return [`assertion_calibration_failed:${check.assertion_id}`];
    if (check.status === "unverified") return [`assertion_calibration_missing:${check.assertion_id}`];
    return [];
  });
  return {
    status,
    required_text_assertions: checks.length,
    calibrated_text_assertions: checks.filter((check) => check.status === "valid").length,
    checks,
    reasons,
  };
}

export function normalizeSampling(raw, { legacyRepeats, determinism }) {
  if (raw === null || raw === undefined) {
    return { repeats: legacyRepeats, pairing: "paired", source: "legacy-determinism" };
  }
  if (typeof raw !== "object" || Array.isArray(raw)) throw new Error("sampling must be an object");
  const unknown = Object.keys(raw).filter((key) => !["repeats", "pairing"].includes(key)).sort();
  if (unknown.length > 0) {
    throw new Error(`sampling contains unsupported fields: ${unknown.join(", ")}`);
  }
  if (!Number.isInteger(raw.repeats) || raw.repeats < 1 || raw.repeats > 10) {
    throw new Error("sampling.repeats must be an integer from 1 to 10");
  }
  const pairing = raw.pairing ?? "paired";
  if (pairing !== "paired") throw new Error("sampling.pairing must be paired");
  if (determinism === "stochastic" && raw.repeats < 3) {
    throw new Error("stochastic evals require at least three sampling repeats");
  }
  return { repeats: raw.repeats, pairing, source: "explicit" };
}

export function assessRuntimeMeasurement({ oracle, sampling, directionDisagreement }) {
  const oracleStatus = String(oracle.status ?? "unverified");
  const reasons = [...(oracle.reasons ?? [])];
  const status = oracleStatus === "invalid"
    ? "invalid"
    : oracleStatus === "valid"
      ? "valid"
      : "unverified";
  return {
    status,
    oracle,
    sampling: {
      status: "valid",
      repeats: sampling.repeats,
      pairing: sampling.pairing,
      source: sampling.source,
      direction_disagreement: directionDisagreement,
    },
    reasons: [...new Set(reasons.map(String))].sort(),
  };
}
